File: server/routes/spotify.ts
import { Hono } from "hono";
import { createMiddleware } from "hono/factory";
import { verify } from "hono/jwt";
import { getAccessKey } from "../db/init";
import { zValidator } from "@hono/zod-validator";
import { z } from "zod";

type Env = {
	Variables: {
		token: string;
	};
};

const getAccessTokenMiddleware = createMiddleware<Env>(async (c, next) => {
	const accessToken = await getAccessKey();
	if (!accessToken) {
		throw new Error("no access token available");
	}
	c.set("token", accessToken);
	await next();
});

const spotifyImageSchema = z.object({
	url: z.string().url(),
	height: z.number().int(),
	width: z.number().int(),
});

const spotifyArtistSchema = z.object({
	external_urls: z
		.object({
			spotify: z.string().url(),
		})
		.optional(),
	followers: z
		.object({
			href: z.null(),
			total: z.number().int(),
		})
		.optional(),
	genres: z.array(z.string()),
	href: z.string().url(),
	images: z.array(spotifyImageSchema),
	name: z.string(),
	popularity: z.number().int(),
	type: z.string().refine((x) => x === "artist"),
	uri: z.string(),
});

const spotifyGetTopArtistSchema = z.object({
	href: z.string().url(),
	next: z.string().url().nullable(),
	previous: z.string().url().nullable(),
	limit: z.number().int(),
	offset: z.number().int(),
	total: z.number().int(),
	items: z.array(spotifyArtistSchema),
});

export const spotifyRoute = new Hono()
	.get("/authorize", async (c) => {
		const { token } = c.req.query();
		try {
			await verify(token, process.env.JWT_SECRET!);
		} catch (err) {
			console.error(err);
			return c.json({ error: "invalid token" } as const, 400);
		}
		const clientID = process.env.SPOTIFY_CLIENT_ID!;
		const redirectURI = process.env.SPOTIFY_REDIRECT_URI!;
		const searchParams = new URLSearchParams({
			client_id: clientID,
			redirect_uri: redirectURI,
			response_type: "code",
			show_dialog: "false",
			scope: "user-top-read",
		});
		return c.redirect(
			`https://accounts.spotify.com/authorize?${searchParams.toString()}`,
		);
	})
	.get("/top/artists", getAccessTokenMiddleware, async (c) => {
		const spotifyAccess = c.var.token;
		const { limit, offset, time_range } = c.req.query();
		const res = await fetch(
			`https://api.spotify.com/v1/me/top/artists?${new URLSearchParams({
				limit,
				offset,
				time_range,
			}).toString()}`,
			{
				method: "GET",
				headers: {
					Authorization: `Bearer ${spotifyAccess}`,
				},
			},
		);
		if (!res.ok) {
			res.json().then((body) => {
				console.error(body);
			});
			throw new Error(`${c.req.url}: ${res.status} ${res.statusText}`);
		}
		const data = await res.json();
		return c.json(spotifyGetTopArtistSchema.parse(data));
	});

File: server/spotify.go
package main

import (
	"encoding/json"
	"fmt"
	"net/http"
	"strconv"
)

type topArtists struct {
	Href    string `json:"href"`
	Limit   int    `json:"limit"`
	Offset  int    `json:"offset"`
	Next    string `json:"next,omitempty"`
	Prev    string `json:"previous,omitempty"`
	Total   int    `json:"total"`
	Artists []struct {
		ExternalURLs struct {
			Spotify string `json:"spotify"`
		} `json:"external_urls"`
		Followers struct {
			Href  string `json:"href,omitempty"`
			Total int    `json:"total"`
		} `json:"followers"`
		Genres     []string       `json:"genres"`
		Href       string         `json:"href"`
		ID         string         `json:"id"`
		Images     []spotifyImage `json:"images"`
		Name       string         `json:"name"`
		Popularity int            `json:"popularity"`
		Type       string         `json:"type"`
		URI        string         `json:"uri"`
	} `json:"items"`
}

type spotifyImage struct {
	URL    string `json:"url"`
	Height int    `json:"height,omitempty"`
	Widhth int    `json:"width,omitempty"`
}

func spotifyTopArtists(w http.ResponseWriter, r *http.Request) {
	search := r.URL.Query()
	limitString := search.Get("limit")
	limit, err := strconv.Atoi(limitString)
	if err != nil {
		respondWithError(w, r, http.StatusInternalServerError, "failed to parse search parameter: limit")
		return
	}
	if limit < 1 || limit > 50 {
		respondWithError(w, r, http.StatusBadRequest, "invalid limit: %d", limit)
		return
	}
	offsetString := search.Get("offset")
	offset, err := strconv.Atoi(offsetString)
	if err != nil {
		respondWithError(w, r, http.StatusInternalServerError, "failed to parse search parameter: offset")
		return
	}
	if offset != 0 {
		respondWithError(w, r, http.StatusBadRequest, "offsets are not supported!")
		return
	}
	timeRange := search.Get("time_range")
	if timeRange != "short_term" && timeRange != "medium_term" && timeRange != "long_term" {
		respondWithError(w, r, http.StatusBadRequest, "invalid time range: %s", timeRange)
		return
	}
	req, err := http.NewRequest(http.MethodGet, "https://api.spotify.com/v1/me/top/artists?"+search.Encode(), nil)
	if err != nil {
		respondWithError(w, r, http.StatusInternalServerError, "erroneous request")
		return
	}
	accessToken, ok := r.Context().Value("spotify-access").(string)
	if !ok {
		respondWithError(w, r, http.StatusInternalServerError, "spotify access token invalid or missing")
		return
	}
	req.Header.Add("Authorization", fmt.Sprintf("Bearer %s", accessToken))
	res, err := http.DefaultClient.Do(req)
	if err != nil {
		respondWithError(w, r, http.StatusInternalServerError, "spotify API request failed")
		return
	} else if res.StatusCode/100 != 2 {
		respondWithError(w, r, res.StatusCode, "spotify API request returned %s", res.Status)
		return
	}
	var body topArtists
	err = json.NewDecoder(res.Body).Decode(&body)
	r.Body.Close()
	if err != nil {
		respondWithError(w, r, http.StatusInternalServerError, "failed to decode response body")
		return
	}
	respondWithJSON(w, http.StatusOK, body)
}
